task_processing/plugins/mesos/execution_framework.py

```diff
@@ -115,31 +119,79 @@ def kill_tasks_stuck_in_staging(self):
-                    ):
-                        log.warning(
-                            'Killing stuck task {id}'.format(id=task_id)
-                        )
-                        self.kill_task(task_id)
-                        self.blacklist_slave(
-                            agent_id=self.task_metadata[task_id].agent_id,
-                            timeout=self.slave_blacklist_timeout_s,
-                        )
-                        get_metric(TASK_STUCK_COUNT).count(1)
+            time_now = time.time()
+            for task_id in self.task_metadata.keys():
+                md = self.task_metadata[task_id]
+
+                eligible_for_termination_at = md.task_state_history[
+                    md.task_state] + self.task_staging_timeout_s
+
+                if time_now < eligible_for_termination_at:
+                    continue
+
+                if md.task_state == 'UNKNOWN':
+                    log.warning('Task {id} has been in unknown state for '
+                                'longer than {timeout}. Re-enqueuing it.'
+                                .format(
+                                    id=task_id,
+                                    timeout=self.task_staging_timeout_s
+                                )
+                                )
+                    self._reenqueue_task(task_id)
+                    get_metric(TASK_LAUNCH_FAILED_COUNT).count(1)
+                    continue
 
+                if md.task_state not in (
+                    'TASK_STAGING',
+                ):
+                    continue
+
+                if time_now > (
+                    md.task_state_history['TASK_STAGING'] +
+                    self.task_staging_timeout_s
+                ):
+                    log.warning(
+                        'Killing stuck task {id}'.format(id=task_id)
+                    )
+                    self.kill_task(task_id)
+                    self.blacklist_slave(
+                        agent_id=self.task_metadata[task_id].agent_id,
+                        timeout=self.slave_blacklist_timeout_s,
+                    )
+                    get_metric(TASK_STUCK_COUNT).count(1)
+
+            self._reconcile_tasks(
+                [Dict({'task_id': task_id}) for
+                    task_id in self.task_metadata if
+                    self.task_metadata[task_id].task_state == 'UNKNOWN']
+            )
             time.sleep(10)
 
+    def _reconcile_tasks(self, tasks_to_reconcile):
+        if time.time() < self._reconcile_tasks_at:
+            return
+
+        log.info('Reconciling following tasks {tasks}'.format(
+            tasks=tasks_to_reconcile
+        ))
+
+        if len(tasks_to_reconcile) > 0:
+            try:
+                self.driver.reconcileTasks(tasks_to_reconcile)
+            except (socket.timeout, Exception):
+                log.warning('Failed to reconcile task status')
+
+        self._reconcile_tasks_at += time.time()
```

Comment:
The increment is task_reconciliation_delay, not time.time().

Reply:
Good catch!